File: golang/conway.go
package main

/*
A simple Conway's Game of Life in Go.

No cycle detection. Infinite grid. No GUI.

Copyright (C) 2014-2025 Youri Ackx under GNU General Public License.
See the LICENSE file and [http://www.gnu.org/licenses/].
*/

// Return all the neighbours of the given cell, living or not.
func neighbours(cell *Cell) *CellSet {
	n := NewCellSet()
	for i := -1; i <= 1; i++ {
		for j := -1; j <= 1; j++ {
			if i != 0 || j != 0 {
				n.Add(Cell{i + cell.X, j + cell.Y})
			}
		}
	}
	return n
}

// Next Moves to the next state
func Next(grid *CellSet) *CellSet {
	newGrid := NewCellSet()
	candidates := NewCellSet()

	// Retrieve all cells to check - living cells and theirs neighbours
	for _, living := range grid.Cells() {
		candidates.Add(living)
		neighbours := neighbours(&living)
		for _, neighbour := range neighbours.Cells() {
			candidates.Add(neighbour)
		}
	}

	// Check each candidate against the game's rules
	for _, cell := range candidates.Cells() {
		livingNeighboursCount := len(neighbours(&cell).Intersect(grid).Cells())
		isAlive := grid.Contains(cell)

		if (isAlive && (livingNeighboursCount == 2 || livingNeighboursCount == 3)) ||
			(!isAlive && livingNeighboursCount == 3) {
			newGrid.Add(cell)
		}
	}

	return newGrid
}

File: haskell/conway.hs
import qualified Data.Set as Set

type Cell = (Int, Int)
type Cells = Set.Set Cell

-- Neighbours of a cell, living or dead
neighbours :: Cell -> Cells
neighbours (x, y) = Set.fromList
    [ (x+dx, y+dy)
    | dx <- [-1..1]
    , dy <- [-1..1]
    , (dx, dy) /= (0, 0) -- Exclude the cell itself
    ]

countLivingNeighbours :: Cell -> Cells -> Int
countLivingNeighbours cell cells = Set.size $ Set.intersection (neighbours cell) cells

isAlive :: Cell -> Cells -> Bool
isAlive cell cells = Set.member cell cells

-- Candidate cells are all living cells and their neighbours
candidates :: Cells -> Cells
candidates grid = Set.unions [grid, Set.unions (Set.map neighbours grid)]

-- Apply game's rules
conway :: Cells -> Cells
conway grid = Set.filter (willLive grid) (candidates grid)
  where
    willLive grid cell
      | isAlive cell grid = n == 2 || n == 3
      | otherwise         = n == 3
      where n = countLivingNeighbours cell grid

displayGrid :: Cells -> IO ()
displayGrid cells = putStrLn $ show (Set.toList cells)

initialState :: Cells
initialState = Set.fromList [(0, 0), (1, 0), (2, 0), (1, 1), (2, 1)]

main :: IO ()
main = do
  displayGrid initialState
  runSimulation initialState
  where
    runSimulation grid 
      | Set.null grid = return ()
      | otherwise = do
          let next = conway grid
          displayGrid next
          runSimulation next
